src/replay_syscall.cc

```diff
@@ -1240,7 +1240,7 @@ static void rep_process_syscall_arch(ReplayTask* t, ReplayTraceStep* step,
       ASSERT(t, t->regs().syscall_result() == trace_regs.syscall_result());
       if (modified_sys == Arch::mprotect) {
         Registers r2 = t->regs();
-        r2.set_arg1(r.arg1());
+        r2.set_orig_arg1(r.arg1());
```

Comment:
I'm not sure about this change.

This code wants to perform the recorded syscall. It seems to me we should have `r.set_arg1(r.orig_arg1())` before line 1234, in case we entered this syscall via `PTRACE_SYSEMU` in which case I guess arg 1 might already be clobbered?

On this particular line here, we're trying to restore the changes made by `fixup_mprotect_growsdown_parameters()`. It seems to me that `fixup_mprotect_growsdown_parameters()` should read `r.orig_arg1()` instead of `r.arg1()`. And here we should also be reading `r.orig_arg1()`, shouldn't we?

Reply:
I believe when I was testing this, I've only seen arg1 being the same as orig_arg1 here within `r`. It seems that all of the changes you suggested depends on

> in case we entered this syscall via PTRACE_SYSEMU in which case I guess arg 1 might already be clobbered?

I'm not familiar with that part. How is PTRACE_SYSEMU used differently from PTRACE_SYSCALL here?

Reply:
`ReplaySession::enter_syscall()` replays up to the next syscall. It usually sets a breakpoint at the syscall instruction and runs forward to that breakpoint, because that's faster, but in some cases it can't do that so it uses `PTRACE_SYSEMU` instead. (See `// Skip this optimization if we can't set the breakpoint`...) So you might only see an issue in the breakpoint-can't-be-set case. You could try modifying the code to always skip that optimization and see what happens to arg 1. I don't know if `PTRACE_SYSEMU` changes arg1 on aarch64.

But maybe it's safer to just change the code to use `orig_arg1()` even when we're not sure that's necessary.

Reply:
I'll try that. Though even if we are using PTRACE_SYSEMU, why would that change the register? I thought we should just get a syscall enter stop?

Reply:
I assume this is what's needed to disable such optimization?

```diff
diff --git a/src/ReplaySession.cc b/src/ReplaySession.cc
index 0a3cc87f..ecfc15fd 100644
--- a/src/ReplaySession.cc
+++ b/src/ReplaySession.cc
@@ -598,13 +598,13 @@ Completion ReplaySession::enter_syscall(ReplayTask* t,
       // in writeable or shared memory, since in those cases it could be
       // overwritten by the tracee. It could even be dynamically generated and
       // not generated yet.
-      if (!syscall_bp_vm &&
-          t->vm()->is_breakpoint_in_private_read_only_memory(
-              syscall_instruction) &&
-          t->vm()->add_breakpoint(syscall_instruction, BKPT_INTERNAL)) {
-        syscall_bp_vm = t->vm();
-        syscall_bp_addr = syscall_instruction;
-      }
+      // if (!syscall_bp_vm &&
+      //     t->vm()->is_breakpoint_in_private_read_only_memory(
+      //         syscall_instruction) &&
+      //     t->vm()->add_breakpoint(syscall_instruction, BKPT_INTERNAL)) {
+      //   syscall_bp_vm = t->vm();
+      //   syscall_bp_addr = syscall_instruction;
+      // }
     }

     if (cont_syscall_boundary(t, constraints) == INCOMPLETE) {
```

Currently replay of julia crashes after I do this, no matter what I do here.

Reply:
Alright, I think we need to document that `ReplaySession::enter_syscall()` completes with the real arg1 in `arg1()` and `orig_arg1()` is undefined.

Reply:
I need to think about these cases a bit more but don't have time right now.

Reply:
No problem, I can remove this one from the pr for now until it’s more clear what we need.

Reply:
I've removed this case for now.